packages/local-relay/src/RelayServer.ts

```diff
@@ -0,0 +1,220 @@
+      return this.rejectWithError(
+        request,
+        401,
+        'Unrecognized connection type, neither authenticated nor connection token provided',
+      );
+    }
+
+    if (isHostConnectionAttempted) {
+      const host: Host = this.hostStore.addOrReplace({
+        id: hostId,
+        displayName,
+        roles: this.getHostRoles(rolesHeader),
+      });
+
+      // Emitted after request.accept()
+      request.on('requestAccepted', (ws: websocket.connection) => {
+        host.connect(ws);
+        console.info(`Accepting host connection ID ${host.id}`);
+      });
+      // "else" == "else if isClientConnectionAttempted", because of the 2 if statements above
+    } else {
+      const host: Host = this.hostStore.get(hostId);
+
+      if (!host) {
+        return this.reject(request, 404, 'Invalid or expired connection token');
+      }
+
+      if (!host.isConnected()) {
+        return this.reject(request, 500, `Host ${host.id} is not connected`);
+      }
+
+      if (!host.canConnectPeer()) {
+        return this.reject(
+          request,
+          403,
+          `Host ${host.id} is already connected to a peer`,
+        );
+      }
+
+      // Emitted after request.accept()
+      request.on('requestAccepted', (ws: websocket.connection) => {
+        try {
+          host.connectPeer(ws);
+        } catch (error) {
+          return this.rejectWithError(request, 500, (error as Error).message);
+        }
+      });
+    }
+
+    // TODO: Add comments
+    return this.accept(request);
+  }
+
+  private onAppGetRoot(_: express.Request, response: express.Response) {
+    response.status(426);
+    response.send('Upgrade Required');
+  }
+
+  private async lookupHosts(_: express.Request, response: express.Response) {
+    const hosts = this.hostStore.listAll();
+
+    response.json({
+      hosts: hosts.map((host) => ({
+        id: host.id,
+        displayName: host.displayName,
+        roles: host.roles,
+        state: host.canConnectPeer() ? 'available' : 'busy',
+      })),
+    });
+  }
+
+  private async createConnectionURL(
+    request: express.Request,
+    response: express.Response,
+  ) {
+    const { id } = request.params;
+    response.setHeader('content-type', 'text/uri-list');
+    response.send(
+      `ws://localhost:${
+        (this.httpServer.address() as net.AddressInfo).port
+      }/${id}\r\n`,
+    );
+  }
+
+  private closeConnection(
+    request: express.Request,
+    response: express.Response,
+  ) {
+    const { connectionId: id } = request.params;
+    this.hostStore.get(id).disconnect();
+    response.status(204).end();
+  }
+
+  private getHostRoles(rolesHeader: string) {
+    if (!rolesHeader) return [];
+
+    const roles = rolesHeader
+      .split(',')
+      .map((role: string) => role.trim().toUpperCase());
+
+    const roleRegex = /^[a-zA-Z0-9_]+$/;
+
+    for (const role of roles) {
+      if (!role.match(roleRegex)) {
+        // Should I throw?
```

Comment:
If you catch in the caller, sure.

Reply:
`https://github.com/Fitbit/developer-bridge/pull/182/commits/ae7fbbeef021e4cc8bf76d1f3c756c13ee8e45f2`